### PR TITLE
sysaudio: rewrite in zig

### LINKED ISSUE #625: sysaudio: rewrite (drop libsoundio dependency)
## Motivation
- more Zig:zap:
- AAudio/Web (and more) backends
- catch bugs
- easily add/drop features

## TODO
- [X] Pulseaudio backend
- [X] Alsa (*unix*) backend
- [X] Dummy backend
- [x] Jack backend
- [x] WASAPI/Asio (*Windows*) backend
- [X] #484
- [ ] CoreAudio (*Apple*) backend
- [ ] AAudio backend (*Android*)
- [ ] Web backend

Currently progress can be followed in [alichraghi/libsoundio](https://github.com/alichraghi/libsoundio)

### PR DESCRIPTION
**Pull request by alichraghi**
- [x] By selecting this checkbox, I agree to license my contributions to this project under the license(s) described in the LICENSE file, and I have the right to do so or have received permission to do so by an employer or client I am producing work for whom has this right.

Closes #625 

### blockers:
- https://github.com/hexops/sdk-linux-x86_64/pull/5
- https://github.com/hexops/sdk-linux-aarch64/pull/3

### CODE REVIEW

**Comment by alichraghi** [comment]
ready for review. CI failure is because system-sdk must be updated and after that piano example i guess

**Comment by kdchambers** [comment]
I'm getting a similar error as the Linux CI, only for Jack. Obviously I don't have an appropriate audio back-end installed (Pipewire), but perhaps this could be caught at comptime and return a nicer error.

As for Context depending on itself, that seems real unfortunately. Maybe just a limitation in Zig that needs to be worked around for now. Similar to https://github.com/ziglang/zig/issues/131

```
const c = @cImport(@cInclude("jack/jack.h"));
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
referenced by:
    Player: libs/sysaudio/src/jack.zig:175:14
    Player: libs/sysaudio/src/jack.zig:169:20
    remaining reference traces hidden; use '-freference-trace' to see all reference traces

/home/keith/github/wrkdir/mach_ali/zig-cache/o/56e0d4bbaaf593154d693621f05d82e3/cimport.h:1:10: error: 'jack/jack.h' file not found
#include <jack/jack.h>
         ^
libs/sysaudio/src/jack.zig:7:21: error: struct 'jack.Context' depends on itself
pub const Context = struct {
                    ^~~~~~
libs/sysaudio/src/jack.zig:7:21: error: struct 'jack.Context' depends on itself
pub const Context = struct {
                    ^~~~~~
libs/sysaudio/src/jack.zig:7:21: error: struct 'jack.Context' depends on itself
pub const Context = struct {
                    ^~~~~~
libs/sysaudio/src/main.zig:104:20: error: struct 'main.Player' depends on itself
pub const Player = struct {
                   ^~~~~~
libs/sysaudio/src/jack.zig:7:21: error: struct 'jack.Context' depends on itself
pub const Context = struct {
                    ^~~~~~
libs/sysaudio/src/jack.zig:7:21: error: struct 'jack.Context' depends on itself
pub const Context = struct {
                    ^~~~~~
libs/sysaudio/src/backends.zig:28:15: error: union 'backends.BackendPlayer__union_4717' depends on itself
    .linux => union(enum) {
              ^~~~~
libs/sysaudio/src/jack.zig:7:21: error: struct 'jack.Context' depends on itself
pub const Context = struct {
                    ^~~~~~
libs/sysaudio/src/backends.zig:28:15: error: union 'backends.BackendPlayer__union_4717' depends on itself
    .linux => union(enum) {
              ^~~~~
libs/sysaudio/src/jack.zig:7:21: error: struct 'jack.Context' depends on itself
pub const Context = struct {
                    ^~~~~~
libs/sysaudio/src/backends.zig:28:15: error: union 'backends.BackendPlayer__union_4717' depends on itself
    .linux => union(enum) {
              ^~~~~
libs/sysaudio/src/backends.zig:28:15: error: union 'backends.BackendPlayer__union_4717' depends on itself
    .linux => union(enum) {
              ^~~~~
libs/sysaudio/src/backends.zig:28:15: error: union 'backends.BackendPlayer__union_4717' depends on itself
    .linux => union(enum) {
```

I also got the following error when cloning recursively:

`fatal: No url found for submodule path 'libs/sysaudio/upstream' in .gitmodule`

Which I fixed by removing the path `git rm --cached libs/sysaudio/upstream`

Seeing as this works for all other targets, if it isn't convenient to test on Linux we could merge this (Or put into a new branch) and I'll fix up these minor issues.

**Comment by alichraghi** [comment]
@kdchambers for now you need to checkout into my SDK PRs i mentioned above.
clone them somewhere and set `MACH_SDK_PATH=<the_sdks_dir>`
also fixed the git error. thanks

File: libs/sysaudio/src/jack.zig [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+            if (c.jack_port_connected_to(port, self.dest_ports[i].ptr) == 1)
+                return false;
+        }
+        return true;
+    }
+
+    pub fn setVolume(self: *Player, vol: f32) !void {
+        _ = self;
+        _ = vol;
+        @panic("incompatible backend");
```
Comment (slimsag) [inline libs/sysaudio/src/jack.zig:240]: Is panic desirable behavior here? Seems like this should be a documented no-op instead maybe.

**Review by slimsag — APPROVED** [review]

**Comment by slimsag** [comment]
Great work on this. I see this as the future so will merge. I think we regress on a few things:

1. macOS is not currently supported
2. Linux requires e.g. pulseaudio and jack to both be installed, since we dynamically link against them. In the future we should make Linux backends [use `dlopen`](https://sourcegraph.com/github.com/ziglang/zig@40ed6ae8469fd599f0524d294f38365c3bb8a825/-/blob/lib/std/dynamic_library.zig?L392-404) so that if not installed we can provide a nicer error / just disable the backend which doesn't exist.
3. WebAudio support is removed for now (but is being worked on.)

We'll need to address these in the near future, but for now this is a great step forward.

**Pull request merged**